module/component/component.go

```diff
@@ -190,79 +232,90 @@ type ComponentManager struct {
-		go func() {
-			select {
-			case err := <-errChan:
-				cancel() // shutdown all workers
-
-				// we propagate the error directly to the parent because a failure in a
-				// worker routine is considered irrecoverable
-				parent.Throw(err)
-			case <-c.done:
-				// Without this additional select, there is a race condition here where the done channel
-				// could be closed right after an irrecoverable error is thrown, so that when the scheduler
-				// yields control back to this goroutine, both channels are available to read from. If this
-				// second case happens to be chosen at random to proceed, then we would return and silently
-				// ignore the error.
-				select {
-				case err := <-errChan:
-					cancel()
-					parent.Throw(err)
-				default:
-				}
+	// Make sure we only start once. atomically check if started is false then set it to true.
+	// If it was not false, panic
+	if !c.started.CAS(false, true) {
+		panic(module.ErrMultipleStartup)
+	}
+
+	ctx, cancel := context.WithCancel(parent)
+	signalerCtx, errChan := irrecoverable.WithSignaler(ctx)
+	go func() {
+		<-ctx.Done()
+		close(c.shutdownSignal)
+	}()
+
+	// launch goroutine to propagate irrecoverable error
+	go func() {
+		// only signal when done channel is closed
+		doneCtx, _ := util.WithDone(context.Background(), c.done)
+		if err := util.WaitError(doneCtx, errChan); err != nil {
+			cancel() // shutdown all workers
+
+			// we propagate the error directly to the parent because a failure in a
+			// worker routine is considered irrecoverable
+			parent.Throw(err)
+		}
+	}()
+
+	// launch workers
+	running := make([]module.ReadyDoneAware, len(c.workers))
+	for i, worker := range c.workers {
+		worker := worker
+		running[i] = worker
+		go worker.run(signalerCtx, c.lookupWorker)
+		if c.serialStart {
+			if err := util.WaitClosed(parent, worker.Ready()); err != nil {
+				break
 			}
-		}()
-
-		var workersReady sync.WaitGroup
-		var workersDone sync.WaitGroup
-		workersReady.Add(len(c.workers))
-		workersDone.Add(len(c.workers))
-
-		// launch workers
-		for _, worker := range c.workers {
-			worker := worker
-			go func() {
-				defer workersDone.Done()
-				var readyOnce sync.Once
-				worker(signalerCtx, func() {
-					readyOnce.Do(func() {
-						workersReady.Done()
-					})
-				})
-			}()
 		}
+	}
 
-		// launch goroutine to close ready channel
-		go c.waitForReady(&workersReady)
+	// launch goroutine to close ready channel
+	go c.waitForReady(running)
 
-		// launch goroutine to close done channel
-		go c.waitForDone(&workersDone)
-	} else {
-		panic(module.ErrMultipleStartup)
+	// launch goroutine to close done channel
+	go c.waitForDone(running)
+}
+
+// lookupWorker returns a ReadyDoneAware interface for the worker specified by ID
+// Prefix id with ! to get a ReadyDoneAware interface for the collection of all workers except the
+// provided id.
+func (c *ComponentManager) lookupWorker(id string) (module.ReadyDoneAware, bool) {
+	if id[0] == '!' {
```

Comment:
I don't quite understand why we need this.

It seems like the main use case is to [filter workers by the ID](https://github.com/onflow/flow-go/blob/petera/consensus-follower-error-handling-multipart2/cmd/scaffold.go#L1012).

I also don't think we should use string as ID here, it's error prone as I mentioned in other comment. 

The reason we need this is because the component worker takes a `LookupFunc` function. If it were taken an interface, which allows caller to pass in different kind of lookup func, it would be more flexible and also safer. For instance:
```
type ComponentWorker func(ctx irrecoverable.SignalerContext, ready ReadyFunc, finder workerFinder)
```
```
AddWorker("main", func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc, finder workerFinder) {
  finder.FindByID(WorkerIDMain)
  finder.FilterByID(WorkerIDMain) 
}

```

But I'm saying I suggest to change into this. I still have question about why we need to do this.

First of all, I have question about the difference between: `Component`, `Worker`, `Module`. Is Worker supposed to check the dependencies of a component are all ready, and then Start the component?

Reply:
After our discussion, I've removed lookup worker in favor of a simpler approach.